host/xtest/pkcs11_1000.c

```diff
@@ -3741,3 +3741,428 @@ static void xtest_pkcs11_test_1016(ADBG_Case_t *c)
+	CK_ATTRIBUTE get_template[] = {
+		{ CKA_CLASS, &g_class, sizeof(CK_OBJECT_CLASS) },
+		{ CKA_KEY_TYPE,	&g_keytype, sizeof(CK_KEY_TYPE) },
+		{ CKA_EXTRACTABLE, &g_extract, sizeof(CK_BBOOL) },
+		{ CKA_SENSITIVE, &g_sensitive, sizeof(CK_BBOOL) },
+		{ CKA_NEVER_EXTRACTABLE, &g_nextract, sizeof(CK_BBOOL) },
+		{ CKA_ALWAYS_SENSITIVE, &g_asensitive, sizeof(CK_BBOOL) },
+		{ CKA_LOCAL, &g_local, sizeof(CK_BBOOL) },
+		{ CKA_VALUE_LEN, &secret_len, sizeof(secret_len) },
+		/*
+		 * CKA_VALUE should remain last attribute in template,
+		 * in this test case as we check the length returned
+		 * from last index of the get_template in this test.
+		 */
+		{ CKA_VALUE, &g_val[0], sizeof(g_val) },
+	};
+	uint32_t idx = ARRAY_SIZE(get_template) - 1;
+	CK_ATTRIBUTE parent_template1[] = {
+		{ CKA_SENSITIVE, &(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) },
+		{ CKA_EXTRACTABLE, &(CK_BBOOL){CK_FALSE}, sizeof(CK_BBOOL) },
+		{ CKA_VALUE_LEN, &(CK_ULONG){16}, sizeof(CK_ULONG) },
+		{ CKA_DERIVE, &(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) },
+	};
+	CK_ATTRIBUTE parent_template2[] = {
+		{ CKA_SENSITIVE, &(CK_BBOOL){CK_FALSE}, sizeof(CK_BBOOL) },
+		{ CKA_EXTRACTABLE, &(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) },
+		{ CKA_VALUE_LEN, &(CK_ULONG){16}, sizeof(CK_ULONG) },
+		{ CKA_ENCRYPT, &(CK_BBOOL){CK_FALSE}, sizeof(CK_BBOOL) },
+		{ CKA_DERIVE, &(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) },
+	};
+	CK_ATTRIBUTE parent_template_wo_derive[] = {
+		{ CKA_SENSITIVE, &(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) },
+		{ CKA_EXTRACTABLE, &(CK_BBOOL){CK_FALSE}, sizeof(CK_BBOOL) },
+		{ CKA_VALUE_LEN, &(CK_ULONG){16}, sizeof(CK_ULONG) },
+	};
+	CK_ATTRIBUTE parent_template_w_enc_der[] = {
+		{ CKA_VALUE_LEN, &(CK_ULONG){16}, sizeof(CK_ULONG) },
+		{ CKA_ENCRYPT, &(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) },
+		{ CKA_DERIVE, &(CK_BBOOL){CK_TRUE}, sizeof(CK_BBOOL) },
+	};
+
+	rv = init_lib_and_find_token_slot(&slot);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		return;
+
+	rv = init_test_token(slot);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto close_lib;
+
+	rv = init_user_test_token(slot);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto close_lib;
+
+	rv = C_OpenSession(slot, session_flags, NULL, 0, &session);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto close_lib;
+
+	/*
+	 * Parent AES Key
+	 * SENSITIVE = TRUE, EXTRACTABLE = FALSE
+	 * ALWAYS_SENSITIVE = TRUE, NEVER_EXTRACTABLE = TRUE
+	 */
+	rv = C_GenerateKey(session, &cktest_aes_keygen_mechanism,
+			   parent_template1, ARRAY_SIZE(parent_template1),
+			   &aes_key1);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto close_session;
+
+	/*
+	 * Parent AES Key
+	 * SENSITIVE = FALSE, EXTRACTABLE = TRUE
+	 * ALWAYS_SENSITIVE = FALSE, NEVER_EXTRACTABLE = FALSE
+	 */
+	rv = C_GenerateKey(session, &cktest_aes_keygen_mechanism,
+			   parent_template2, ARRAY_SIZE(parent_template2),
+			   &aes_key2);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto close_session;
+
+	Do_ADBG_BeginSubCase(c, "Derive Generic secret - AES-ECB Mechanism");
+
+	/*
+	 * Use AES key 1 as Parent key
```

Comment:
as base key?

Reply:
I changed it to Parent key throughout to maintain consistency. Do you want me to change it to "Base Key" everywhere ?

Reply:
I let you choose -- base key term is just used in C_DeriveKey().